### ISSUE TITLE
[DoesNotReturn] warning is only reported in nullable-enabled context

### ISSUE THREAD

**Issue by cston** [dotnet/csharplang#5231]
Warning CS8763 should be reported for `F1()` below:
```csharp
using System.Diagnostics.CodeAnalysis;

class Program
{
    [DoesNotReturn] static void F1() { } // no warning

#nullable enable
    [DoesNotReturn] static void F2() { } // warning CS8763
}

namespace System.Diagnostics.CodeAnalysis
{
    public class DoesNotReturnAttribute : Attribute { }
}
```

**Comment by jcouv** [comment]
Moved the issue to csharplang since it's a language design question.

**Comment by jcouv** [comment]
Added topic to LDM backlog ([when convenient](https://github.com/dotnet/csharplang/blob/main/meetings/2021/README.md#schedule-when-convenient))

**Comment by 333fred** [comment]
@jcouv like Rikki said, I don't understand what we're trying to ask here. This is the expected behavior of all nullable flow analysis attributes: what else are you expecting to see?

**Comment by jcouv** [comment]
@333fred Yes, the current behavior is by current design. The ask is to change this design ;-)

The proposed change and expectation are described in OP. Unlike other nullability attributes, this one doesn't have anything to do with nullability.
If we explore this, there are some other additions we should discuss too (effect on definite assignment). For example, I think this warning would fall out from this proposal:
```
F1();
int i = 0; // warning CS0162: Unreachable code detected
```

**Comment by RikkiGibson** [comment]
This seems about equally sound as definite assignment of `out` parameters. There's no way to guarantee that a method with an `out` parameter from metadata never reads the argument before assigning it, or that the method assigns it at all.

```cs
void M()
{
    int x;
    SomeLibrary.M(out x);
    x.ToString(); // could still be uninitialized?
}
```

**Comment by 333fred** [comment]
> Unlike other nullability attributes, this one doesn't have anything to do with nullability

It _only_ has to do with nullability. No other part of the language or compiler pays attention to any of these attributes. It does not affect definite assignment in any fashion. I guess I'm not sure what has changed in the past 2 years that we feel we need to revisit this again.

**Comment by 333fred** [comment]
@jcouv did you have a response to my question above?

**Comment by jcouv** [comment]
@333fred  It's a question for Chuck who's raising the issue. Personally, I am sympathetic to the proposal and simply having more experience with the attributes is a sufficient reason to revisit.

**Comment by cston** [comment]
> It _only_ has to do with nullability.

The attribute indicates only that the method does not return which seems independent of nullability, so it's surprising the corresponding warning is only reported in a `#nullable enable` context.

**Comment by Youssef1313** [comment]
I can see this both ways.


```csharp
if (x is null)
{
    ADoesNotReturnMethod();
}

_= x.ToString() // no warning. This is the original intent of the attribute.
```

However, from another perspective, it's just a "DoesNotReturn" where I want the compiler to warn me if it's used while the method **does** return. So whether or not the compiler is using it for nullability analysis is just one side of it.

**Comment by RikkiGibson** [comment]
Are we comfortable introducing a new error in implementations when DoesNotReturn is applied, but the method does return? That's what would be necessary to meet "definite assignment" standards, but the shipped behavior is only to warn.

**Comment by CyrusNajmabadi** [comment]
> Discussed in LDM 10/13/2021: in short, keep as-is

In long: we are more interested in investing our efforts into a more powerful, generalized, solution like `never`/`bottom` types: https://github.com/dotnet/csharplang/issues/538

**Comment by 333fred** [comment]
Notes: https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-10-13.md#revisiting-doesnotreturn
